dkg/dkg.go

```diff
@@ -48,6 +48,55 @@ type Config struct {
 	TestSigning bool
 }
 
+// exchanger is responsible for exchanging partial signatures between peers on libp2p.
+type exchanger struct {
+	setChan chan core.ParSignedDataSet
+	sigex   *parsigex.ParSigEx
+}
+
+func newExchanger(tcpNode host.Host, peerIdx int, peers []peer.ID) exchanger {
+	ex := exchanger{
+		sigex:   parsigex.NewParSigEx(tcpNode, peerIdx, peers),
+		setChan: make(chan core.ParSignedDataSet, len(peers)),
+	}
+
+	// Wiring core workflow components
+	ex.sigex.Subscribe(ex.getSets)
+
+	return ex
+}
+
+func (e *exchanger) exchange(ctx context.Context, set core.ParSignedDataSet) ([]core.ParSignedDataSet, error) {
+	err := e.sigex.Broadcast(ctx, core.Duty{}, set)
+	if err != nil {
+		return nil, err
+	}
+
+	var sets []core.ParSignedDataSet
+	sets = append(sets, set)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, err
+		case peerSet := <-e.setChan:
+			sets = append(sets, peerSet)
+		}
+
+		// We are done when we have ParSignedDataSet from all the peers including the current one
+		if len(sets) == cap(e.setChan) {
```

Comment:
I would just compare against length of peers rather. Also remember that peers might be send duplicate messages by mistake (network retries, bugs, byzantine), so I would dedup

Reply:
yeah right